src/stores/UserProfilesStore.ts

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { logger } from "matrix-js-sdk/src/logger";
+import { IMatrixProfile, MatrixClient, MatrixEvent, RoomMember, RoomMemberEvent } from "matrix-js-sdk/src/matrix";
+
+import { LruCache } from "../utils/LruCache";
+
+const cacheSize = 500;
+
+type StoreProfileValue = IMatrixProfile | undefined | null;
+
+/**
+ * This store provides cached access to user profiles.
+ * Listens for membership events and invalidates the cache for a profile on update with different profile values.
+ */
+export class UserProfilesStore {
+    private profiles = new LruCache<string, IMatrixProfile | null>(cacheSize);
+    private knownProfiles = new LruCache<string, IMatrixProfile | null>(cacheSize);
+
+    public constructor(private client?: MatrixClient) {
```

Comment:
Why is the client optional?

Reply:
The client is optional in SDKContext.

Reply:
Sure, but what sense does constructing a UserProfileStore without a client make? Especially given it is a singleton. Without a client it'll never cull stale data. Along with the race condition listed below that sounds likely to introduce bugs.